lib/wibox/widget/template.lua

```diff
@@ -134,44 +158,171 @@ function template:update(args)
+--     }
+--     for _, signal in pairs(signals) do
+--         self._private.widget:connect_signal(signal, function(...)
+--             self:emit_signal(signal, ...)
+--         end)
+--     end
+
+    self:emit_signal("widget::redraw_needed")
+    self:emit_signal("property::template", widget_template)
+end
+
+--- Set a property on one or more template sub-widget instances.
+--
+-- This method allows to set a value at any time on any of the sub widget of
+-- the template. To use this, you can set, or document a set of `ids` which
+-- your template support. These are usually referred as "roles" across the
+-- other APIs.
+--
+--@DOC_wibox_widget_template_set_property_existing_EXAMPLE@
+--
+-- It is also possible to take this one step further and apply a property
+-- to the entire sub-widget tree. This allows users to implement their own
+-- template even if it doesn't use the same "roles" as the default one:
+--
+--@DOC_wibox_widget_template_set_property_custom_EXAMPLE@
+--
+-- @method set_property
+-- @tparam string property The property name.
+-- @param value The property value.
+-- @tparam[opt=nil] nil|string|table ids A sub-widget `id` or list of `id`s. Use
+--  `nil` for "all sub widgets".
+-- @tparam[opt=false] boolean recursive Also set the property on all sub-widgets.
+-- @noreturn
+-- @see bind_property
+function template:set_property(property, value, ids, recursive)
+    recursive = recursive or false
+
+    local widgets
+    local target = self._private.widget
+
+    -- Lazy load later.
+    if not target then
+        table.insert(self._private.connections, {
+            src_obj  = nil,
+            src_prop = nil,
+            once     = true,
+            apply    = function()
+                self:set_property(property, value, ids, recursive)
+            end,
+        })
+        return
     end
 
-    self._private.template = widget
-    self._private.widget = widget_instance
+    if ids then
+        widgets = {}
+        for _, id in ipairs(type(ids) == "string" and {ids} or ids) do
+            for _, widget in ipairs(target:get_children_by_id(id)) do
+                table.insert(widgets, widget)
+            end
+        end
+    end
 
-    -- We need to connect to these signals to actually redraw the template
-    -- widget when its child needs to.
-    local signals = {
-        "widget::redraw_needed",
-        "widget::layout_changed",
-    }
-    for _, signal in pairs(signals) do
-        self._private.widget:connect_signal(signal, function(...)
-            self:emit_signal(signal, ...)
-        end)
+    widgets = widgets or {target}
+
+    local apply_property
+    apply_property = function(wdgs)
+        for _, widget in ipairs(wdgs) do
+            if widget["set_"..property] then
+                widget["set_"..property](widget, value)
+            end
+
+            if recursive and widget.get_children then
+                apply_property(widget:get_children())
+            end
```

Comment:
I never liked this approach of just blindly assigning a property to the whole tree. And I don't see any merit in it over explicitly setting role `id`s where needed.

It's inefficient compared to the `by_id` caches, especially without any checks on how big the tree is that you're recursing into.

The fact that some properties will "magically" be available on all widgets while others need an `id` will lead to confusion (and already does with things like `icon_role` vs `awful.widget.clienticon` in the tasklist examples).

With this being a public API assigning arbitrary properties to arbitrary widgets, chances rise that not every `:set_foo()` expects the same `foo`, i.e. not every widget (especially when taking 3rd-party libraries into account) might expect the same value for the same property name.
E.g. if there was a widget template that used a `fg` or `bg` property to set a color based on a value, then you wouldn't want to apply that to an entire widget tree where there might be `wibox.container.backgrounds` being overwritten.
That wasn't so big of a deal until now, as the only internal implementations of this were `tag` for the taglist and `client` for the tasklist, where you could reasonably assume that any widget that has a `:set_client()` actually wants an `awful.client` instance. But now, there is no way to reason about what a template and its usage may be.

It's also weird that when `ids` is set, it only recurses into their children. I can't think of a use case where I'd set a role on a widget and then also need that role on all of its child widgets. Given our separation of layout/container/content widgets, widgets that have child widgets don't have content to display of their own.

So while it's just an optional parameter, giving the user a choice, I think it's a choice they should never take. And while we could just reword the method docs to heavily discourage using the option, simply removing it would prevent the confusion and make the API simpler.